# ML-03 — Frame Your Lane as an ML Task

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane as an ML task (type)

*Classification, clustering, ranking, or scoring — which one, and why?*

Lane 2 is a ranking and scoring task. Each eligible page receives a priority score, and the highest-scored pages form the review queue. A classification model could estimate the probability that a page matches the decline proxy, and that probability could help determine its priority score. However, the final output is an ordered ranking of pages.

## 2. Target or proxy

*What would you predict? Where does that label come from — observed outcome or a defined rule?*

My target is a decline proxy defined as `trend_direction == "down"`. A page receives this label when its impressions in the most recent 30 days are more than 20% below the previous 30 days. This is a defined rule based on measured impression change. However, it does not show that a page requires a refresh or that editing the page would improve its performance.

## 3. Success metric

*One metric you can defend. What number means 'good'?*

My main metric is Precision@50. It measures the proportion of pages labeled by the decline proxy among the 50 pages ranked highest for review. This matches the decision because the example assumes that the team has 50 refresh slots per month. I will consider a Precision@50 of at least 0.70 good. This means that at least 35 of the 50 pages selected for review match the decline proxy.

## 4. The unit of analysis, as a real dataframe

*Load your lane's slice and show it: one row = one what?*

In [1]:
import pandas as pd

df = pd.read_csv("../../data/raw/content_refresh_anonymized.csv")
visible = df["impressions_90d"] >= 100
columns = ["content_id", "content_type", "content_age_days", "days_since_last_update", "impressions_90d", "clicks_90d", "avg_position"]

lane_df = df.loc[visible, columns]
lane_df["is_declining_proxy"] = df.loc[visible, "trend_direction"].eq("down")

print(f"Starter rows: {len(df):,}")
print(f"Eligible review pages: {len(lane_df):,}")
lane_df.head()

Starter rows: 30,000
Eligible review pages: 22,006


,content_id,content_type,content_age_days,days_since_last_update,impressions_90d,clicks_90d,avg_position,is_declining_proxy
0,content_304f48230142,keyword article,187,20,3803,29,10.6,True
1,content_a1fb4e703a9e,keyword article,445,25,15320,7,20.3,True
2,content_9aa793d4d895,keyword article,141,20,12581,11,36.5,True
3,content_331d6c4de07b,keyword article,463,22,11751,58,6.2,False
4,content_d99b7a2d90ca,keyword article,263,14,19140,24,44.0,True


In the starter dataset, one row = one pseudonymized content page. Its search and engagement measurements cover a trailing 90-day period. The dataset contains 30,000 unique pages, of which 22,006 meet the eligibility threshold of at least 100 impressions. This threshold removes pages with less evidence, but it doesn't mean that pages below this threshold have no value.

## 5. Why ML beats a fixed rule here

*What makes the pattern too messy for an if-statement?*

A fixed rule is not enough because page priority depends on many signals at the same time, including impressions, clicks, search position, CTR, engagement, content age, and freshness. These signals also interact. Covering all their possible combinations with hand-written rules would take a long time and produce a complicated system that is difficult to maintain. Machine learning can learn these patterns from the examples, combine the signals into one score, and rank pages more effectively for editorial review. The starter results support this: the hand-written rule achieved 0.24 Precision@50, while the model achieved 0.74.

## Self-check

Before you submit, confirm each line honestly:

- [X] Every section above is filled — markdown thinking AND the code that backs it
- [X] The notebook runs top to bottom with no errors (Runtime → Run all)
- [X] No client names, URLs, or private queries anywhere
- [X] My claims use careful words: observed, measured, directional, decision-support
- [X] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.